core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -178,82 +189,105 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
-                buildNodeDecisions(nodesToAllocate, explain));
+            throttled = true;
+        }
+
+        Map<String, NodeAllocationResult> nodeResults = null;
+        if (explain) {
+            nodeResults = buildNodeDecisions(nodesToAllocate, inSyncAllocationIds);
+        }
+        if (allocation.hasPendingAsyncFetch()) {
+            return AllocateUnassignedDecision.no(AllocationStatus.FETCHING_SHARD_DATA, nodeResults);
+        } else if (node != null) {
+            return AllocateUnassignedDecision.yes(node, allocationId, nodeResults, false);
+        } else if (throttled) {
+            return AllocateUnassignedDecision.throttle(nodeResults);
+        } else {
+            return AllocateUnassignedDecision.no(AllocationStatus.DECIDERS_NO, nodeResults, true);
         }
     }
 
     /**
      * Builds a map of nodes to the corresponding allocation decisions for those nodes.
      */
-    private static Map<String, Decision> buildNodeDecisions(NodesToAllocate nodesToAllocate, boolean explain) {
-        if (explain == false) {
-            // not in explain mode, no need to return node level decisions
-            return null;
-        }
-        Map<String, Decision> nodeDecisions = new LinkedHashMap<>();
-        for (final DecidedNode decidedNode : nodesToAllocate.yesNodeShards) {
-            nodeDecisions.put(decidedNode.nodeShardState.getNode().getId(), decidedNode.decision);
-        }
-        for (final DecidedNode decidedNode : nodesToAllocate.throttleNodeShards) {
-            nodeDecisions.put(decidedNode.nodeShardState.getNode().getId(), decidedNode.decision);
+    private static Map<String, NodeAllocationResult> buildNodeDecisions(NodesToAllocate nodesToAllocate, Set<String> inSyncAllocationIds) {
+        return Stream.of(nodesToAllocate.yesNodeShards, nodesToAllocate.throttleNodeShards, nodesToAllocate.noNodeShards)
+                   .flatMap(Collection::stream)
+                   .collect(Collectors.toMap(dnode -> dnode.nodeShardState.getNode().getId(),
+                       dnode -> new NodeAllocationResult(dnode.nodeShardState.getNode(),
+                                                         shardStoreInfo(dnode.nodeShardState, inSyncAllocationIds),
+                                                         dnode.decision),
+                       (dnode1, dnode2) ->  { throw new IllegalStateException(String.format(Locale.ROOT, "Duplicate key %s", dnode1)); },
+                       LinkedHashMap::new));
+    }
+
+    private static ShardStore shardStoreInfo(NodeGatewayStartedShards nodeShardState, Set<String> inSyncAllocationIds) {
+        final Exception storeErr = nodeShardState.storeException();
+        final StoreReadability storeReadability;
+        if (storeErr != null) {
+            Throwable unwrapped = ExceptionsHelper.unwrapCause(storeErr);
+            if (unwrapped instanceof CorruptIndexException) {
+                storeReadability = StoreReadability.CORRUPT;
```

Comment:
I wonder if we need an enum for this as we already have the exception at hand in ShardStore.